File: src/agent/codex-appserver/client-pool.ts
import { statSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { log } from '../../core/logger';
import { AppServerClient, JsonRpcError } from './app-server-client';
import { resolveCodexBin } from './locate';

/**
 * 常驻 app-server 进程（M-2，依据 research/08 实测）：
 *
 *  - **utility client**：thread/list、thread/read、model/list、account/read 强刷
 *    这类元数据 RPC 原本每次各付一整套 spawn+initialize（~500ms）。现在共享一个
 *    懒创建、**出错即重建**的常驻进程；这些 RPC 的 cwd 都是请求参数（thread/list
 *    的 cwd 是过滤条件），与进程 cwd 无关，所以进程 cwd 给中性目录即可。
 *
 *  - **预热池（容量 1）**：冷路径 spawn→initialize≈495ms + 首个 thread/start≈1640ms
 *    （MCP server 启动），同进程第二个 thread/start 仅 ≈64ms。池里预先放一个
 *    spawn+initialize+ephemeral thread/start（触发 MCP 预热、不落盘）完毕的进程，
 *    新会话直接取走（真正的 thread/start|resume 的 cwd/sandbox/权限都是 thread 级
 *    参数，与进程无关——README 与 research/08 已核）；取走后异步补位。
 *
 * 生命周期：会话客户端取走后归 sessions/orchestrator 管（shutdown 会 close）；
 * 池中待命进程与 utility client 归本模块管——bot 的 shutdown 不知道这里，所以挂
 * 一个 process 'exit' 钩子按 pid 兜底 SIGKILL（run.ts 优雅退出最终 process.exit(0)，
 * 必经此钩子），另导出 {@link shutdownResidentClients} 供测试/未来显式接线。
 */

/** 中性目录：池进程与 utility 进程的 spawn cwd。真实会话的 cwd 是 thread/start
 * 的参数，与进程 cwd 无关；预热的 ephemeral thread 也用它，绝不沾项目目录。 */
const NEUTRAL_CWD = tmpdir();

/** initialize 握手死线：utility/预热进程连不上必须可恢复（卡死的 create 单飞
 * Promise 会饿死后续所有 acquire），超时即 close（SIGKILL）重来。 */
const CONNECT_TIMEOUT_MS = 15_000;

/** 预热 thread/start 死线：实测 ~1.6s（MCP 启动），放宽兜异常环境。 */
const PREWARM_TIMEOUT_MS = 60_000;

/** utility RPC 默认死线。原 listThreads/listModels 连 close 都到不了的「挂死到天荒
 * 地老」从此有界；超时即杀进程重建（等价旧短命进程的 finally close 语义）。 */
const DEFAULT_UTILITY_TIMEOUT_MS = 30_000;

function withDeadline<T>(p: Promise<T>, ms: number, label: string): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    const t = setTimeout(() => reject(new Error(`${label} timed out after ${ms}ms`)), ms);
    p.then(
      (v) => {
        clearTimeout(t);
        resolve(v);
      },
      (e) => {
        clearTimeout(t);
        reject(e);
      },
    );
  });
}

// ── 常驻进程登记 + 退出兜底 ───────────────────────────────────────────

const residents = new Set<AppServerClient>();
let exitHookInstalled = false;

function track(client: AppServerClient): void {
  residents.add(client);
  if (exitHookInstalled) return;
  exitHookInstalled = true;
  // 'exit' 里没有事件循环，close() 的优雅等待跑不完——直接按 pid SIGKILL。
  // 会话客户端不在 residents 里（orchestrator.shutdown 已优雅 close 它们）。
  process.once('exit', () => {
    for (const c of residents) {
      const pid = c.pid;
      if (pid === undefined) continue;
      try {
        process.kill(pid, 'SIGKILL');
      } catch {
        // already dead
      }
    }
    residents.clear();
  });
}

function untrack(client: AppServerClient): void {
  residents.delete(client);
}

// ── utility client（懒创建、出错即重建） ──────────────────────────────

let utility: { client: AppServerClient; bin: string } | null = null;
let utilityCreating: Promise<unknown> | null = null;

async function acquireUtility(bin: string): Promise<AppServerClient> {
  // 单飞：先等在途的创建落定，再看槽位（绝不并发 spawn 两个 utility）。
  while (utilityCreating) await utilityCreating.catch(() => undefined);
  const cur = utility;
  if (cur && !cur.client.exited && cur.bin === bin) return cur.client;
  if (cur) {
    // 进程死了 / codex 二进制换了位置——重建
    utility = null;
    untrack(cur.client);
    void cur.client.close().catch(() => undefined);
  }
  const create = (async () => {
    const client = new AppServerClient({ bin, cwd: NEUTRAL_CWD, clientName: 'feishu-codex-bridge-utility' });
    // 先登记再连接：在途的 connect 撞上进程退出时，exit 钩子也能按 pid 兜底。
    track(client);
    try {
      await withDeadline(client.connect(), CONNECT_TIMEOUT_MS, 'utility connect');
    } catch (err) {
      untrack(client);
      void client.close().catch(() => undefined);
      throw err;
    }
    // 没人消费常驻客户端的通知流（它不跑 thread）——永远排空，防 AsyncQueue
    // 无界增长；进程退出时流自然结束。
    void (async () => {
      for await (const n of client.stream()) void n;
    })();
    utility = { client, bin };
    log.info('agent', 'utility-up', { pid: client.pid ?? null });
    return client;
  })();
  utilityCreating = create.catch(() => undefined);
  try {
    return await create;
  } finally {
    utilityCreating = null;
  }
}

function discardUtility(client: AppServerClient): void {
  if (utility?.client === client) utility = null;
  untrack(client);
  void client.close().catch(() => undefined);
}

/**
 * 在共享 utility client 上发一个 JSON-RPC 请求（懒连接）。连接失败 / 进程死亡 /
 * 超时都**丢弃当前进程**，下一次调用拿全新进程——「出错即重建」。超时丢弃即
 * close（SIGKILL），保住旧短命进程对 wedged codex 的「杀进程」恢复语义
 * （readHistory 的 20s 死线靠这个不留孤儿）。
 *
 * 例外：应用层 JSON-RPC error 应答（如 thread/read 指向已被 codex 删除的会话）
 * 说明进程健康，原样上抛、进程保留——否则一个调用方的业务错误会 SIGKILL 共享
 * 进程，failAllPending 把并发在飞的其他请求（account/read 强刷、thread/list）
 * 全部打挂，正是本仓库最忌讳的「时灵时不灵」。
 */
export async function utilityRequest<T = unknown>(
  method: string,
  params?: unknown,
  opts?: { timeoutMs?: number },
): Promise<T> {
  const bin = resolveCodexBin();
  if (!bin) throw new Error('codex CLI not found (set CODEX_BIN or install @openai/codex)');
  const client = await acquireUtility(bin);
  try {
    return await withDeadline(
      client.request<T>(method, params),
      opts?.timeoutMs ?? DEFAULT_UTILITY_TIMEOUT_MS,
      `utility ${method}`,
    );
  } catch (err) {
    if (!(err instanceof JsonRpcError)) discardUtility(client);
    throw err;
  }
}

// ── 预热池（容量 1） ──────────────────────────────────────────────────

interface WarmEntry {
  client: AppServerClient;
  bin: string;
  /** bin 文件指纹（mtime+size）——取用时复验，防池中进程跨 codex 升级版本错位 */
  fingerprint: string | null;
}

let warm: WarmEntry | null = null;
let warming: Promise<void> | null = null;

/** bin 文件指纹 = 「codex 版本」的取用时探活代理。statSync 是微秒级且跟随符号
 * 链接，能在取用热进程的瞬间发现「codex 已原地升级/被替换」；版本字符串探测
 * （codex --version ~320ms）放在取用路径会吃掉预热的全部收益，而 locate 的版本
 * 缓存恰恰看不见原地升级——指纹比版本号更严也更便宜。 */
function binFingerprint(bin: string): string | null {
  try {
    const s = statSync(bin);
    return `${s.mtimeMs}:${s.size}`;
  } catch {
    return null;
  }
}

/**
 * 取走预热进程（探活：进程活着 + 还是同一个 codex 二进制）。同步执行，并发
 * spawn 不可能双取。失活/版本错位即弃置（refill 会按新二进制补位），返回 null
 * 走冷路径。
 */
export function takeWarmClient(bin: string): AppServerClient | null {
  const entry = warm;
  if (!entry) return null;
  warm = null;
  untrack(entry.client);
  if (entry.client.exited) return null; // 池中阵亡——冷路径接管
  if (entry.bin !== bin || entry.fingerprint !== binFingerprint(bin)) {
    // codex 升级/换位：版本错位的待命进程不能再接新会话
    log.info('agent', 'prewarm-stale', { bin });
    void entry.client.close().catch(() => undefined);
    return null;
  }
  // 预热期间缓冲的通知（MCP 启动进度、ephemeral thread/started…）属于预热
  // 线程，清空——绝不能漏进真实会话的事件流。
  entry.client.clearNotifications();
  log.info('agent', 'prewarm-hit', { pid: entry.client.pid ?? null });
  return entry.client;
}

/**
 * 异步补位（取走即补 / 扑空也补，容量恒 1）。预热 = spawn + initialize +
 * ephemeral thread/start（触发 MCP 启动这 ~1.6s 的大头；ephemeral 不落盘、
 * 弃置无痕）。失败只记日志——预热是纯优化，绝不影响请求路径。
 * 返回在途的补位 Promise（永不 reject），调用方通常 fire-and-forget。
 */
export function refillWarmPool(): Promise<void> {
  if (warm || warming) return warming ?? Promise.resolve();
  warming = (async () => {
    const bin = resolveCodexBin();
    if (!bin) return;
    // 指纹在 spawn 之前取——若 spawn 与升级赛跑，取用时的复验自会兜住。
    const fingerprint = binFingerprint(bin);
    // clientName 用默认值：这个进程取走后就是真实会话的进程（compliance log 同名）。
    const client = new AppServerClient({ bin, cwd: NEUTRAL_CWD });
    // 先登记再预热：补位进行中 daemon 退出时，exit 钩子也能按 pid 兜底。
    track(client);
    try {
      await withDeadline(client.connect(), CONNECT_TIMEOUT_MS, 'prewarm connect');
      await withDeadline(
        client.request('thread/start', {
          cwd: NEUTRAL_CWD,
          ephemeral: true,
          approvalPolicy: 'never',
          sandbox: 'read-only',
        }),
        PREWARM_TIMEOUT_MS,
        'prewarm thread/start',
      );
      warm = { client, bin, fingerprint };
      log.info('agent', 'prewarm-ready', { pid: client.pid ?? null });
    } catch (err) {
      log.fail('agent', err, { phase: 'prewarm' });
      untrack(client);
      void client.close().catch(() => undefined);
    }
  })().finally(() => {
    warming = null;
  });
  return warming;
}

// ── 清理 ──────────────────────────────────────────────────────────────

/** 优雅关掉全部常驻进程（utility + 预热槽，含在途补位）。daemon 路径由 process
 * 'exit' 钩子兜底；这个出口给测试与未来的显式接线用。 */
export async function shutdownResidentClients(): Promise<void> {
  // 等在途的创建/补位落定，否则它们会在我们清完之后才把进程放进槽位
  while (utilityCreating) await utilityCreating.catch(() => undefined);
  if (warming) await warming.catch(() => undefined);
  const targets: AppServerClient[] = [];
  if (utility) {
    targets.push(utility.client);
    utility = null;
  }
  if (warm) {
    targets.push(warm.client);
    warm = null;
  }
  for (const c of targets) untrack(c);
  await Promise.allSettled(targets.map((c) => c.close()));
}

File: internal/agent/codex/clientpool.go
package codex

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"sync"
	"time"
)

// clientpool.go —— 常驻 codex 进程池（utility + 预热池容量 1，对齐 TS client-pool）。
//
//   - utility client：thread/list、thread/read、model/list、account/read 等元数据 RPC
//     共享一个懒创建、出错即重建的常驻进程（避免每次 spawn+initialize ~500ms）。
//   - 预热池容量 1：预先 spawn+initialize+ephemeral thread/start（触发 MCP ~1.6s 大头），
//     新会话直接取走（thread/start 从 ~2.1s 降到 ~64ms）；取走后异步补位。
//
// 关键不变量：JsonRpcError 不杀进程（进程健康）；超时/传输层失败才 discard+SIGKILL。

const (
	connectTimeout        = 15 * time.Second
	prewarmTimeout        = 60 * time.Second
	defaultUtilityTimeout = 30 * time.Second
)

// Pool 常驻进程池。
type Pool struct {
	bin        func() string // 默认 ResolveCodexBin(false)
	neutralCWD string

	mu           sync.Mutex
	util         *utilEntry    // utility client
	utilCreating chan struct{} // 单飞（非 nil=在途创建）
	warm         *warmEntry    // 预热槽
	warming      bool          // 预热单飞

	residents map[*AppServerClient]struct{} // exit 兜底 SIGKILL 集合
	resMu     sync.Mutex
}

type utilEntry struct {
	client *AppServerClient
	bin    string
}

type warmEntry struct {
	client *AppServerClient
	bin    string
	fp     string
}

// defaultPool 进程级默认池（生产用）。
var defaultPool = newPool()

func newPool() *Pool {
	return &Pool{
		bin:        func() string { return ResolveCodexBin(false) },
		neutralCWD: os.TempDir(),
		residents:  map[*AppServerClient]struct{}{},
	}
}

// UtilityRequest 在共享 utility client 上发 RPC（默认池）。
func UtilityRequest(ctx context.Context, method string, params any, timeout time.Duration) (json.RawMessage, error) {
	return defaultPool.UtilityRequest(ctx, method, params, timeout)
}

// TakeWarmClient 取走预热进程（默认池）。
func TakeWarmClient(bin string) *AppServerClient { return defaultPool.TakeWarmClient(bin) }

// RefillWarmPool 异步补位预热池（默认池）。
func RefillWarmPool() { defaultPool.RefillWarmPool() }

// ShutdownResidentClients 关闭全部常驻进程（默认池）。
func ShutdownResidentClients() { defaultPool.Shutdown() }

// UtilityRequest 在共享 utility client 上发 RPC。
// 连接失败/进程死亡/超时 → 丢弃进程重建；JsonRpcError → 原样上抛、进程保留。
func (p *Pool) UtilityRequest(ctx context.Context, method string, params any, timeout time.Duration) (json.RawMessage, error) {
	if timeout <= 0 {
		timeout = defaultUtilityTimeout
	}
	bin := p.bin()
	if bin == "" {
		return nil, errors.New("codex CLI not found (set CODEX_BIN or install @openai/codex)")
	}
	client, err := p.acquireUtility(ctx, bin)
	if err != nil {
		return nil, err
	}
	rctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	res, err := client.Request(rctx, method, params)
	if err != nil && !AsJsonRpcError(err) {
		p.discardUtility(client)
	}
	return res, err
}

// acquireUtility 单飞获取 utility client：在途创建者把后来者阻塞在 chan 上。
func (p *Pool) acquireUtility(ctx context.Context, bin string) (*AppServerClient, error) {
	p.mu.Lock()
	for p.utilCreating != nil {
		ch := p.utilCreating
		p.mu.Unlock()
		select {
		case <-ch:
		case <-ctx.Done():
			return nil, ctx.Err()
		}
		p.mu.Lock()
	}
	// 复用：进程活 + 同一 bin
	if p.util != nil && !p.util.client.Exited() && p.util.bin == bin {
		c := p.util.client
		p.mu.Unlock()
		return c, nil
	}
	// 死了 / bin 换了 → 丢弃旧
	if p.util != nil {
		old := p.util.client
		p.util = nil
		p.mu.Unlock()
		p.untrack(old)
		old.Close(0)
		p.mu.Lock()
	}
	// 单飞创建
	creating := make(chan struct{})
	p.utilCreating = creating
	p.mu.Unlock()

	client, err := p.connectResident(ctx, bin, "feishu-codex-bridge-utility")

	p.mu.Lock()
	p.utilCreating = nil
	close(creating)
	if err != nil {
		p.mu.Unlock()
		return nil, err
	}
	p.util = &utilEntry{client: client, bin: bin}
	p.mu.Unlock()
	go drainStream(client) // utility 不跑 thread，永远排空通知流防无界增长
	return client, nil
}

func (p *Pool) discardUtility(c *AppServerClient) {
	p.mu.Lock()
	if p.util != nil && p.util.client == c {
		p.util = nil
	}
	p.mu.Unlock()
	p.untrack(c)
	c.Close(0)
}

func (p *Pool) connectResident(ctx context.Context, bin, name string) (*AppServerClient, error) {
	client := NewAppServerClient(AppServerClientOptions{Bin: bin, Cwd: p.neutralCWD, ClientName: name})
	p.track(client)
	cctx, cancel := context.WithTimeout(ctx, connectTimeout)
	defer cancel()
	if err := client.Connect(cctx); err != nil {
		p.untrack(client)
		client.Close(0)
		return nil, err
	}
	return client, nil
}

// TakeWarmClient 取走预热进程（探活：进程活 + 同一 bin 指纹）。失活/错位返回 nil。
func (p *Pool) TakeWarmClient(bin string) *AppServerClient {
	p.mu.Lock()
	entry := p.warm
	p.warm = nil
	p.mu.Unlock()
	if entry == nil {
		return nil
	}
	p.untrack(entry.client)
	if entry.client.Exited() {
		return nil
	}
	if entry.bin != bin || entry.fp != binFingerprint(bin) {
		entry.client.Close(0)
		return nil
	}
	// 清空预热期缓冲的通知（MCP 进度、ephemeral thread/started）——绝不漏进真实会话。
	entry.client.ClearNotifications()
	return entry.client
}

// RefillWarmPool 异步补位（容量恒 1）。失败只记日志——预热是纯优化。
func (p *Pool) RefillWarmPool() {
	p.mu.Lock()
	if p.warm != nil || p.warming {
		p.mu.Unlock()
		return
	}
	p.warming = true
	p.mu.Unlock()

	go func() {
		defer func() {
			p.mu.Lock()
			p.warming = false
			p.mu.Unlock()
		}()
		bin := p.bin()
		if bin == "" {
			return
		}
		fp := binFingerprint(bin)
		client, err := p.connectResident(context.Background(), bin, "")
		if err != nil {
			return
		}
		pctx, cancel := context.WithTimeout(context.Background(), prewarmTimeout)
		defer cancel()
		_, err = client.Request(pctx, "thread/start", map[string]any{
			"cwd":            p.neutralCWD,
			"ephemeral":      true,
			"approvalPolicy": "never",
			"sandbox":        "read-only",
		})
		if err != nil {
			p.untrack(client)
			client.Close(0)
			return
		}
		p.mu.Lock()
		p.warm = &warmEntry{client: client, bin: bin, fp: fp}
		p.mu.Unlock()
	}()
}

// Shutdown 关闭全部常驻进程。
func (p *Pool) Shutdown() {
	p.mu.Lock()
	var targets []*AppServerClient
	if p.util != nil {
		targets = append(targets, p.util.client)
		p.util = nil
	}
	if p.warm != nil {
		targets = append(targets, p.warm.client)
		p.warm = nil
	}
	p.mu.Unlock()
	for _, c := range targets {
		p.untrack(c)
		c.Close(time.Second)
	}
}

// ── 辅助 ────────────────────────────────────────────────────────

// binFingerprint bin 文件指纹（mtime+size），取用时复验防 codex 原地升级错位。
func binFingerprint(bin string) string {
	fi, err := os.Stat(bin)
	if err != nil {
		return ""
	}
	return fmt.Sprintf("%d:%d", fi.ModTime().UnixNano(), fi.Size())
}

func (p *Pool) track(c *AppServerClient) {
	p.resMu.Lock()
	p.residents[c] = struct{}{}
	p.resMu.Unlock()
}

func (p *Pool) untrack(c *AppServerClient) {
	p.resMu.Lock()
	delete(p.residents, c)
	p.resMu.Unlock()
}

func drainStream(c *AppServerClient) {
	for range c.Stream() {
	}
}
